frame/support/src/traits/error.rs

```diff
@@ -0,0 +1,54 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Traits for describing and constraining pallet error types.
+use codec::{Decode, Encode, OptionBool};
+use sp_std::marker::PhantomData;
+
+/// Trait indicating that the implementing type is going to be included as a field in a variant of
+/// the `#[pallet::error]` enum type.
+///
+/// ## Notes
+/// The pallet error enum has a maximum encoded size as defined by
+/// [`frame_support::MAX_PALLET_ERROR_ENCODED_SIZE`]. If the pallet error type exceeds this size
+/// limit, a static assertion during compilation will fail. The compilation error will be in the
+/// format of `error[E0080]: evaluation of constant value failed` due to the usage of
+/// [`static_assertions::const_assert`].
+pub trait PalletError: Encode + Decode {
```

Comment:
Why not just also require `MaxEncodedLen` trait here?

Reply:
I could actually, in fact maybe I should even put the associated const in `MaxEncodedLen` and remove this trait altogether.

Reply:
Wait no, I do need the `Decode` bound for safe measure, but that can be simply derived on the pallet Error enum.

Reply:
Okay, I can derive `MaxEncodedLen`, but I'll need to tinker its bounds so that it doesn't require `T` or `I` to also implement `MaxEncodedLen`, and at this point it feels a bit weird to add a lot more parsing and data storage just to adjust the bounds.

I think it will be worth the effort when I can integrate `PalletError` with `MaxEncodedLen`, but alas that isn't quite possible right now because `BoundedVec`'s upper bounds cannot be used in a const eval context, so we'll need to wait for a couple of Rust features to be implemented first before the integration happens.